### PR TITLE
Feat/restore preamble

### PR DESCRIPTION
**Pull request by willemneal**
### What

Replaces #892 
Depends on #942 

### Why

[TODO: Why this change is being made. Include any context required to understand the why.]

### Known limitations

[TODO or N/A]

### CODE REVIEW

File: cmd/soroban-rpc/internal/test/cli_test.go [hunk 1]
```diff
@@ -78,11 +177,30 @@ func runCLICommand(t *testing.T, cmd string) *icmd.Result {
...
 	)
-	return icmd.RunCmd(c)
+	return c
+}
+
+func runCLICommand(t *testing.T, cmd string) *icmd.Result {
+	return icmd.RunCmd(cliCmd(t, cmd))
+}
+
+func testAccount(t *testing.T, id uint32) string {
```
Comment (2opremio) [inline cmd/soroban-rpc/internal/test/cli_test.go:187]:
```suggestion
func verifyAccountExists(t *testing.T, id uint32) string {
```

File: cmd/soroban-rpc/internal/test/cli_test.go [hunk 2]
```diff
@@ -78,11 +177,30 @@ func runCLICommand(t *testing.T, cmd string) *icmd.Result {
...
+	return strings.Trim(runSuccessfulCLICmd(t, fmt.Sprintf("config identity address --hd-path %d", id)), "\n")
+}
+
+func testContractID(t *testing.T, id uint32) [32]byte {
```
Comment (2opremio) [inline cmd/soroban-rpc/internal/test/cli_test.go:191]:
```suggestion
func getTestContractIDFromAccountAndSalt(t *testing.T, id uint32) [32]byte {
```

File: cmd/soroban-rpc/internal/test/cli_test.go [hunk 3]
```diff
@@ -62,13 +66,108 @@ func TestCLIContractDeployAndInvoke(t *testing.T) {
...
+	count = runSuccessfulCLICmd(t, fmt.Sprintf("contract invoke --id %s -- inc", contractID))
+	require.Equal(t, "2", count)
+}
+
+func symToScVal(s xdr.ScSymbol) xdr.ScVal {
+	res, _ := xdr.NewScVal(xdr.ScValTypeScvSymbol, s)
+	return res
+}
+
+func waitForKey(t *testing.T, test *Test, contractID [32]byte, scValKey xdr.ScVal) {
```
Comment (2opremio) [inline cmd/soroban-rpc/internal/test/cli_test.go:84]:
```suggestion
func waitForKeyToExpire(t *testing.T, test *Test, contractID [32]byte, scValKey xdr.ScVal) {
```

File: cmd/soroban-rpc/internal/test/cli_test.go [hunk 4]
```diff
@@ -62,13 +66,108 @@ func TestCLIContractDeployAndInvoke(t *testing.T) {
 	require.Contains(t, output, `["Hello","world"]`)
 }
 
+func TestCLISimulateTransactionBumpAndRestoreFootprint(t *testing.T) {
```
Comment (2opremio) [inline cmd/soroban-rpc/internal/test/cli_test.go:69]: This test doesn't really bump or restore, could you pick a more appropriate name?

Comment (2opremio) [reply cmd/soroban-rpc/internal/test/cli_test.go:69]: Ah, great!

File: cmd/soroban-cli/src/rpc/mod.rs [hunk 5]
```diff
@@ -626,22 +627,26 @@ soroban config identity fund {address} --helper-url <url>"#
...
             .iter()
             .map(DiagnosticEvent::from_xdr_base64)
             .collect::<Result<Vec<_>, _>>()?;
-
-        Ok((assemble(tx, &sim_response)?, events))
+        Ok((
+            assemble(tx, &sim_response)?,
+            sim_response
+                .restore_preamble
+                .filter(|p| !p.transaction_data.is_empty()),
```
Comment (Shaptic) [inline cmd/soroban-cli/src/rpc/mod.rs:647]: maybe I'm misunderstanding the code here, but I don't think it should ever be empty if the `restore_preamble` is present in the rpc response

File: cmd/soroban-rpc/internal/test/cli_test.go [hunk 6]
```diff
@@ -50,25 +51,37 @@ func outputsContractIDInLastLine(t *testing.T, output string) {
...
+	count = runSuccessfulCLICmd(t, fmt.Sprintf("contract invoke --id %s -- inc", strkeyContractID))
+	require.Equal(t, "3", count)
+}
 func runSuccessfulCLICmd(t *testing.T, cmd string) string {
 	res := runCLICommand(t, cmd)
-	require.NoError(t, res.Error, fmt.Sprintf("stderr:\n%s\nstdout:\n%s\n", res.Stderr(), res.Stdout()))
-	return res.Stdout()
+	stdout, stderr := res.Stdout(), res.Stderr()
+	require.NoError(t, res.Error, fmt.Sprintf("stderr:\n%s\nstdout:\n%s\n", stderr, stdout))
+	println(fmt.Sprintf("stderr:\n%s\nstdout:\n-------\n%s\n-----\n", stderr, stdout))
```
Comment (2opremio) [inline cmd/soroban-rpc/internal/test/cli_test.go:80]:
```suggestion
	fmt.Printf("stderr:\n%s\nstdout:\n-------\n%s\n-----\n", stderr, stdout)
```

File: cmd/soroban-cli/src/rpc/mod.rs (hunk 5 above)
Comment (willemneal) [reply cmd/soroban-cli/src/rpc/mod.rs:647]: Peeking into the output from other RPC tests it seems that an empty `TransactionData` is being sent when it should be ignored.

File: cmd/soroban-rpc/internal/test/cli_test.go [hunk 7]
```diff
@@ -50,25 +51,37 @@ func outputsContractIDInLastLine(t *testing.T, output string) {
...
 }
 
+func TestCLISimulateTransactionRestoreFromPreambleInFootprint(t *testing.T) {
+	NewCLITest(t)
+	strkeyContractID := runSuccessfulCLICmd(t, "contract deploy --salt=0 --wasm "+helloWorldContractPath)
+	count := runSuccessfulCLICmd(t, fmt.Sprintf("contract invoke --id %s -- inc", strkeyContractID))
+	require.Equal(t, "1", count)
+	count = runSuccessfulCLICmd(t, fmt.Sprintf("contract invoke --id %s -- inc", strkeyContractID))
+	require.Equal(t, "2", count)
+	time.Sleep(time.Second * 60)
```
Comment (2opremio) [inline cmd/soroban-rpc/internal/test/cli_test.go:72]: Can you wait for the ledger entry to expire? (with Go)

File: cmd/soroban-rpc/internal/test/cli_test.go [hunk 8]
```diff
@@ -78,11 +91,37 @@ func runCLICommand(t *testing.T, cmd string) *icmd.Result {
...
+}
+
+func getAccountFromID(t *testing.T, id uint32) string {
+	return strings.Trim(runSuccessfulCLICmd(t, fmt.Sprintf("config identity address --hd-path %d", id)), "\n")
 }
 
+// func getTestContractIDFromAccountAndSalt(t *testing.T, id uint32) [32]byte {
+// 	return getContractID(t, getAccountFromID(t, id), testSalt, StandaloneNetworkPassphrase)
+// }
+
```
Comment (2opremio) [inline cmd/soroban-rpc/internal/test/cli_test.go:115]:
```suggestion
```

File: cmd/soroban-rpc/internal/test/cli_test.go (hunk 7 above)
Comment (willemneal) [reply cmd/soroban-rpc/internal/test/cli_test.go:72]: I've tried several different ways and all result it it only reading account keys and not the ledger key.

File: cmd/soroban-rpc/internal/test/cli_test.go [hunk 9]
```diff
@@ -78,11 +91,37 @@ func runCLICommand(t *testing.T, cmd string) *icmd.Result {
...
+const MILLION string = "1000000"
+
 func NewCLITest(t *testing.T) *Test {
 	test := NewTest(t)
+	fundAccount(t, test, getAccountFromID(t, 0), MILLION)
```
Comment (2opremio) [inline cmd/soroban-rpc/internal/test/cli_test.go:120]:
```suggestion
func NewCLITest(t *testing.T) *Test {
	test := NewTest(t)
	fundAccount(t, test, getAccountFromID(t, 0), "1000000")
```

File: cmd/soroban-rpc/internal/test/cli_test.go (hunk 7 above)
Comment (2opremio) [reply cmd/soroban-rpc/internal/test/cli_test.go:72]: Let me try

File: cmd/soroban-cli/src/rpc/mod.rs (hunk 5 above)
Comment (willemneal) [reply cmd/soroban-cli/src/rpc/mod.rs]: Rebased to your commit and removed it! Thanks!

**Review by 2opremio — APPROVED** [review]

**Comment by Shaptic** [comment]
@2opremio @willemneal this PR isn't a release blocker, right?

**Comment by 2opremio** [comment]
Sorry, I was off. Not a blocked, but definitely very nice to have.

**Pull request merged**